File: src/Test/java/com/nishant/algorithms/DataStructures/Trees/BinarySearchTrees/RedBlackTreeTest.java
package com.nishant.algorithms.DataStructures.Trees.BinarySearchTrees;

import com.nishant.algorithms.DataStructures.Trees.BinarySearchTrees.RedBlackTree.RedBlackTree;
import org.junit.Before;
import org.junit.Test;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

public class RedBlackTreeTest {

  private RedBlackTree<Integer> tree;

  @Before
  public void setup() {
    tree = new RedBlackTree<>();
  }

  @Test
  public void insertTest() {
    assertFalse(tree.contains(5));
    tree.insert(5);
    assertTrue(tree.contains(5));

    for (int i = 0; i < 100; i++) {
      tree.insert((int) (Math.random() * 100));
      assertTrue(tree.isValid());
    }
  }

  @Test
  public void deleteTest() {
    tree.insert(5);
    assertTrue(tree.contains(5));
    tree.delete(5);
    assertFalse(tree.contains(5));

    int[] values = new int[100];

    for (int i = 0; i < values.length; i++) {
      values[i] = (int) (Math.random() * 100);
      tree.insert(values[i]);
    }

    for (int i : values) {
      tree.delete(i);
      assertTrue(tree.isValid());
    }
  }
}

File: src/test/cpp/datastructures/trees/binarysearchtrees/redblacktree/red_black_tree_test.cpp
#include <gtest/gtest.h>
#include <cmath>
#include <vector>
#include <set>

#include "main/cpp/datastructures/trees/binarysearchtrees/redblacktree/RedBlackTree.h"

#include "main/cpp/my_utils.h"

TEST(RedBlackTreeTest, RBTreeTest) {
  std::vector<int> values = my_utils::createRandVector(50, 0, 200);
  std::set<int> valueSet;
  for (int i = 0; i < values.size(); i++)
    valueSet.insert(values[i]);
  
  RedBlackTree<int> tree;
  for (int i = 0; i < values.size(); i++) {
    tree.insert(values[i]);
    ASSERT_TRUE(tree.isValid());
    ASSERT_LE(tree.getHeight(), 2*std::log2(tree.numNodes()+1));
  }
  ASSERT_EQ(tree.numNodes(), 50);
  for (int i = 0; i < values.size(); i++) {
    ASSERT_TRUE(tree.contains(values[i]));
  }
  for (int i = 0; i <= 200; i++) {
    ASSERT_EQ(tree.contains(i), valueSet.count(i));
  }
  for (int i = 0; i < values.size(); i++) {
    ASSERT_TRUE(tree.remove(values[i]));
    ASSERT_TRUE(tree.isValid());
  }
  ASSERT_EQ(tree.numNodes(), 0);
}

int main(int argc, char **argv) {
  ::testing::InitGoogleTest(&argc, argv);
  
  return RUN_ALL_TESTS();
}
